### DISCUSSION TITLE
Add package api documentation

### DISCUSSION THREAD

**Discussion by DanieliusDev** [npm/feedback#852] [Website]
Add api documentation for packages published to npm. This would make package developers not need to worry about creating their own documentation websites. This would most likely be implemented in https://docs.npmjs.org. Typescript **must** be supported, and it would be nice if we could build docs locally using a command like `npm build-docs`. Here is how i would like to implement this:

## /packages/{package}/{version}

Get documentation for a package that is **not** scoped. If the package is owned by an organisation or user, the sidebar should include a dropdown with the packages the organisation or user owns linking them to their documentation .

## /packages/@{scope}

Get documentation for packages owned by a user or organisation. This page would list out the packages the organisation or user owns linking them to their documentation.

## /packages/@{scope}/{package}/{version}

Get documentation for a scoped package. Somewhere on the sidebar, there should be a dropdown for packages that are owned by the user or organisation linking them to their own documentation (useful for multi package projects).

---

`version` should be optional in paths and redirect to the `latest` version.

I would gladly help implement this, and if there are any questions, feel free to ask.

**Comment by ljharb** [comment]
Why would a maintainer want their package to automatically have documentation they don't control?

**Reply by DanieliusDev** [reply]
I have api documentation in mind. Like the classes, functions, and type definitions in the packages. Other kinds of guides and stuff should require the developer to create a custom website.

**Reply by ljharb** [reply]
Right, but that's something individual packages may choose. For most of my packages, which are simple, i actively don't *want* such a heavy documentation website - and for the ones which are not, I want the freedom to update them without having to publish another update.

**Reply by DanieliusDev** [reply]
> the

Why wouldn't you want advanced documentation made for you with no extra work? I think you're turning down a really great feature some programming languages with package managers already have. This could also be good for packages who have poor api documentation, and people can just use the npm one.

**Reply by ljharb** [reply]
Good documentation needs to be hand-written; I would never want autogenerated prose to represent my project.

**Reply by DanieliusDev** [reply]
> Good documentation needs to be hand-written; I would never want autogenerated prose to represent my project.

You're missing the actual point of the feature. Api documentation is different from hand-written ones. Hand-written ones can sometimes be harder to maintain (if you're manually documenting api, aka exported items in a package). Im aware there are tools for this, but this would be cool for all npm packages. Hand-written documentation would be good for guides and cli packages.

**Reply by ljharb** [reply]
I'm saying all documentation, especially API documentation, should be hand-written, and if it's hard to manually document the API, that's a great indicator that the API is too complex.

My opinion here doesn't have to be universally held! However, it contradicts the belief that autogenerated documentation is a universally desired thing.
